clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -1011,14 +1029,27 @@ llvm::PreserveCheriTags CodeGenTypes::copyShouldPreserveTags(const Expr *E) {
   QualType Ty = E->IgnoreParenImpCasts()->getType();
   if (Ty->isAnyPointerType())
     Ty = Ty->getPointeeType();
-  // TODO: Find the underlying VarDecl to improve diagnostics
-  const VarDecl *UnderlyingVar = nullptr;
-  // TODO: this assert may be overly aggressive
-  assert((!UnderlyingVar || UnderlyingVar->getType() == Ty) &&
-         "Passed wrong VarDecl?");
-  // If we have an underlying VarDecl, we can assume that the dynamic type of
-  // the object is known and can perform more detailed analysis.
-  return copyShouldPreserveTagsForPointee(Ty, UnderlyingVar != nullptr);
+  bool EffectiveTypeKnown = false;
+  const VarDecl *UnderlyingVar = findUnderlyingVarDecl(E);
+  if (UnderlyingVar) {
+    QualType VarTy = UnderlyingVar->getType();
+    assert(!VarTy->isIncompleteType() && "Unexpected incomplete type");
+    if (VarTy->isReferenceType()) {
+      // If the variable declaration is a C++ reference we can assume that the
```

Comment:
Not sure how useful it is since they could still be a subclass that does contain capabilities. Would need further analysis to check the size of the copy...